### PR TITLE
API: Add `npt.NDArray`, a runtime-subscriptable alias for `np.ndarray`

### PR DESCRIPTION
**Pull request by BvB93**
This PR introduces a runtime-subscriptable alias for `np.ndarray[Any, np.dtype[~Scalar]]` by the name of `npt.NDArray`.

Follow up on https://github.com/numpy/numpy/pull/17719, which made the initial attempt at introducing aforementioned type alias before being deferred to a later PR.

Motivation for the introduction of this type alias is two-fold:
* It allows one to annotate an array with a given dtype during runtime, _e.g._ the same role `typing.Sequence` fulfilled for `collections.abc.Sequence` prior to python 3.9. While runetime-subsription is already, more or less, possible to with `from __future__ import annotations`, the latter only helps with annotations, and not the creation of type aliases. All in all `npt.NDArray` would add a big convenience factor.
* It provides a compact alias for an otherwise rather verbose expression, a convenience that was already extensive utilized in the numpy stubs via its private predecessor (https://github.com/numpy/numpy/commit/c332c6aec120a96f9f08bf330938dd6e0eb2e7d2). Note this alias cannot be used for custom dtypes that are not based on `np.generic` (see [NEP 42](https://numpy.org/neps/nep-0042-new-dtypes.html)), which will either have to define their own aliases or stick to the complete form.

Future Work
------------
Currently I can think of two other (potential) candidates:
* An alias for `np.dtype[~Scalar]`. I'm not 100% sure if the big dtype update (NEP 40-42) already has plans on adding a `dtype.__class_getitem__` method or not, so adding a subscriptable alias to `numpy.typing` might be redudant. @seberg do you have any comments this?
* An alias for `np.ndarray[~Shape, np.dtype[~Scalar]]`. I'd say this would be worthwhile to introduce, once we have shape-typing sorted out (https://github.com/numpy/numpy/issues/16544), that is. 
 There is an open question of what would be the best name for such type-alias. `npt.Array` perhaps? To distinguish it from the arbitrary shaped `npt.NDarray` introduced in this PR.

Implementation
----------------
Fortunately [PEP 585](https://www.python.org/dev/peps/pep-0585/) has provided us with all the tools we need for constructing what are effectively subscriptable wrappers: [`types.GenericAlias`](https://docs.python.org/3/library/types.html#types.GenericAlias). Aforementioned class is designed such that (nearly) all operations wrap around the underlying type, _e.g._ `np.ndarray` behaves virtually the same as `types.GenericAlias(np.ndarray, ())`.

The only bad news is that `GenericAlias` is exclusive to python >= 3.9 (and written in C), hence the introduction of a (python-based) backport in this PR. The backport can of course be removed once we've dropped support for 3.8 at some point in the future.

Examples
----------
``` python
>>> import numpy as np
>>> import numpy.typing as npt

>>> print(npt.NDArray)
numpy.ndarray[typing.Any, numpy.dtype[~ScalarType]]

>>> print(npt.NDArray[np.float64])
numpy.ndarray[typing.Any, numpy.dtype[numpy.float64]]

>>> NDArrayInt = npt.NDArray[np.int_]
>>> a: NDArrayInt = np.arange(10)

>>> def func(a: npt.ArrayLike) -> npt.NDArray[Any]:
...     return np.array(a)
```

### CODE REVIEW

**Comment by seberg** [comment]
I am happy to just add `__class_getitem__`, I have not done it yet, simply because it wasn't one of the "big things to worry about".  I like the syntax (and it seems very reasonable for typing).

@BvB93 if it helps you, I can add this over the weekend, we ping the mailing list about comments (I doubt anyone will be bothered) and then you have it.

File: numpy/typing/_generic_alias.py [hunk 1]
```diff
@@ -0,0 +1,208 @@
+from __future__ import annotations
+
+import sys
+import types
+from typing import (
+    Any,
+    Callable,
```
Comment (s-mrb) [inline numpy/typing/_generic_alias.py:7]: hope `Callable` was not intended to be used anywhere in code.

**Comment by BvB93** [comment]
> I am happy to just add `__class_getitem__`, I have not done it yet, simply because it wasn't one of the "big things to worry about". I like the syntax (and it seems very reasonable for typing).

That would be nice, yes; it's definitely one of the things on my wish lists.
Personally I wouldn't mind waiting for 1.22 either, so no hurries.

**Comment by seberg** [comment]
@BvB93 what is the typical behviour?  Since `np.dtype[scalar]` would return subclasses of `np.dtype`, `__class_getitem__` should refuse to work on subclasses (e.g. `np.dtype[scalar][scalar]` makes no sense) right?

**Comment by BvB93** [comment]
> @BvB93 what is the typical behviour? Since `np.dtype[scalar]` would return subclasses of `np.dtype`, `__class_getitem__` should refuse to work on subclasses (e.g. `np.dtype[scalar][scalar]` makes no sense) right?

I would actually propose to just stick same semantics as cpython used in PEP 585 for the builtin types (https://github.com/python/cpython/pull/18239). For example, with the likes of `builtins.list` their pure python implementation would be equivalent the code below:

``` python
import types

def __class_getitem__(cls, args):
    return types.GenericAlias(cls, args)
```

Or a bit more fancy and and some basic error checking (i.e. this will raise for `dtype[Any, Any]` and `dtype[()]`):

``` python
import types

Scalar = TypeVar("Scalar")
_ALIAS = types.GenericAlias(cls, (Scalar,))

def __class_getitem__(cls, args):
    return _ALIAS[args]  # Let `GenericAlias` handle the error checking
```

This would have a few consequences though:
* `types.GenericAlias` is only available for python >= 3.9. So we could either limit `__class_getitem__`, or for example use the backport introduced in this PR. I'd be ok with either option.
* The return type would, during runtime, be a wrapped version of the original class (so not an actual subclass).

**Comment by seberg** [comment]
> The return type would, during runtime, be a wrapped version of the original class (so not an actual subclass).

Wait, but that sounds pretty much useless at run-time? I thought this would be useful semantics to get the actual subclass.

**Comment by shoyer** [comment]
No objections from me -- this looks very handy!

**Comment by seberg** [comment]
@BvB93 no, its not really any work at all, because I *already* have that. The point is that this is what happens if you type `np.array()`:

```
np.array([scalar])
```
must already be able to effectively do:
```
DType = np.dtype[scalar]
descriptor = DType.get_descriptor_from_pyobject(scalar)
result = np.empty(1, dtype=descriptor)
result[0] = scalar
```
In other words: I already have a function that does `DType = np.dtype[scalar]` in NumPy.  (Of course there could be DTypes without scalars in the future, but in that case the user would have to use that DTypes name directly in any case.)

**Comment by BvB93** [comment]
FYI the discussion on `dtype.__class_getitem__` from this weeks community meeting seems to have stirred towards the `types.GenericAlias`-based approach (xref [python/cpython#18239](https://github.com/python/cpython/pull/18239)). As a reminder, such an enhancement would still be beyond the scope of this particular PR and reserved for future work.

Secondly, I'd like to point the attention of (potential) reviewers to the snippet below, as it would be useful to sort out _some_ of these details already (most importantly: would the names make sense?).

> Future Work
> -------------
> * An alias for `np.ndarray[~Shape, np.dtype[~Scalar]]`. I'd say this would be worthwhile to introduce, once we have shape-typing sorted out (#16544), that is.
There is an open question of what would be the best name for such type-alias. `npt.Array` perhaps? To distinguish it from the arbitrary shaped `npt.NDarray` introduced in this PR.

**Comment by charris** [comment]
Thanks Bas.

**Pull request merged**